impeller/entity/geometry/circle_tessellator.cc

```diff
@@ -0,0 +1,108 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/impeller/entity/geometry/circle_tessellator.h"
+
+#include "flutter/fml/logging.h"
+
+namespace impeller {
+
+std::vector<Trig> CircleTessellator::trigs_[MAX_DIVISIONS_ + 1];
+
+size_t CircleTessellator::ComputeQuadrantDivisions(Scalar pixel_radius) {
+  // Note: these values are approximated based on the values returned from
+  // the decomposition of 4 cubics performed by Path::CreatePolyline.
+  if (pixel_radius < 1.0) {
+    return 1;
+  }
+  if (pixel_radius < 2.0) {
+    return 2;
+  }
+  if (pixel_radius < 12.0) {
+    return 6;
+  }
+  if (pixel_radius <= 36.0) {
+    return 9;
+  }
+  pixel_radius /= 4;
+  if (pixel_radius > (MAX_DIVISIONS_ - 1)) {
+    return MAX_DIVISIONS_;
+  }
+  return static_cast<int>(ceil(pixel_radius));
+}
+
+const std::vector<Trig>& CircleTessellator::GetTrigForDivisions(
+    size_t divisions) {
+  FML_DCHECK(divisions > 0 && divisions <= MAX_DIVISIONS_);
+  std::vector<Trig>& trigs = trigs_[divisions];
+
+  if (trigs.empty()) {
+    double angle_scale = kPiOver2 / divisions;
```

Comment:
If you know the number of divisions consider reserving the size.

Reply:
Just to note, though, that this is only ever executed once per app lifetime per one of 35 possible division counts. But, I added the reserve to help keep over-allocating...

Reply:
None of the tessellation methods (libtess or convex tessellator) , or even the render pass itself are threadsafe. If you add locking at this level it won't make the rest of the code thread safe, and it will add overhead that isn't strictly necessary.

Consider documenting that this class isn't thread safe.

Reply:
But unlike those other cases, this is a static method. The objects may not care about thread safety because they are only ever used from one thread (by design), but can we say the same thing about the class itself and its static methods?

Reply:
Oh I see. Rather than having each circle tessellatior instance reach into shared static storage, can you have this tessellatior live with the convex tessellator and libtess tessellator? Then it could store this computed data as instance state.

Reply:
Done. I also took the time to use a more direct computation of the number of divisions based on geometry/trigonometry and left a comment in the code documenting the equation.